### DISCUSSION TITLE
Implementing a feature-complete web application?

### DISCUSSION THREAD

**Discussion by metaperl** [Kitware/trame#318] [Q&A]
I am just getting back to Sebastien Jourdain who replied to my comment on his medium post awhile back:
https://medium.com/@metaperl/kindly-see-my-survey-guide-on-pure-python-web-application-development-https-metaperl-github-io-p-9213b9bab401

In reviewing Trame, it is astounding. My question is: if you wanted a fully-featured web application with sessions, authentication and authorization, how would you do it? Currently, to list a framework as Class A it must do that [and a few other things](https://metaperl.github.io/pure-python-web-development/practical-guide.html#requirements-of-a-class-a-system). 

Can Trame easily be embedded in a Flask/Django app? It certainly can be deployed [in a wide number of other places](https://kitware.github.io/trame/docs/deploy-pypi.html).

**Comment by jourdain** [comment]
Thanks for looking at trame and evaluating it. 

Before diving into your question, let me provide some background around trame. 

First of all, trame was aimed to create desktop like application with a web front-end. Therefore, what you call a session was aimed to be a dedicated process when using trame. That way we have complete isolation between different users. In fact it was even meant to run on HPC where you want several machines working via MPI to serve a single user. 

But to support Jupyter, we nicely isolated our trame server (aka session) to be an async task. Therefore, you can easily start/stop sessions from a single python environment that support async tasks. 

Regarding your flask and django question, I guess I don't know for sure how asyncio friendly they are. But if they are, then very little would be needed to wrap the start/stop of a trame server as a task within the current python runtime. 

Regarding authentication, you have 2 approaches. One could be managed at the infrastructure of the trame server start/stop logic using the framework capabilities (flask, django, fastapi). Or you can do it within trame using the keycloak library in which trame integrates. 

HTH,

Seb
